libtenzir/include/tenzir/concept/parseable/numeric/integral.hpp

```diff
@@ -36,92 +37,91 @@ template <class T,
-    return parse_pos(f, l, a);
-  }
-
-  template <class Iterator, class Attribute>
-  static bool dispatch(Iterator& f, const Iterator& l, Attribute& a) {
-    if constexpr (std::is_signed_v<T>)
-      return parse_signed(f, l, a);
-    else
-      return parse_unsigned(f, l, a);
+  constexpr static auto is_digit(const char c) -> bool {
+    constexpr auto in_range
+      = [](const char c, const char lower, const char upper) {
+          return c >= lower and c <= upper;
+        };
+    return in_range(c, '0', '0' + std::min(Radix, 9))
+           or (Radix > 10
+               and (in_range(c, 'a', 'a' + Radix - 10)
+                    or in_range(c, 'A', 'A' + Radix - 10)));
   }
 
   template <class Iterator, class Attribute>
   bool parse(Iterator& f, const Iterator& l, Attribute& a) const {
-    if (f == l)
+    if (f == l) {
       return false;
+    }
+    constexpr static bool is_signed = std::is_signed_v<attribute>;
+    const auto first_is_sign = *f == '-' or *f == '+';
+    if (first_is_sign and not is_signed) {
+      return false;
+    }
+    // an array large enough to hold a value of the given type
+    // and a potential sign character
+    // Its a raw array due to better rendering in debuggers
+    char data[MaxDigits + is_signed]{};
+    auto end = std::begin(data);
+    int take_chars = 0;
     auto save = f;
-    if (dispatch(f, l, a))
-      return true;
-    f = save;
+    while (true) {
+      auto c = *f;
+      // take only digits
+      if (not is_digit(c)) {
+        if constexpr (not is_signed) {
+          break;
+        }
+        if (take_chars > 0 or not first_is_sign) {
+          break;
+        }
+      }
+      *end = c;
+      ++take_chars;
+      ++end;
+      ++f;
+      // take at most MaxDigits chars + a potential minus
+      if (take_chars >= MaxDigits + int{first_is_sign}) {
+        break;
+      }
+      // take at most as many chars as the array can hold
+      if (end == std::end(data)) {
+        break;
+      }
+      // take at most as many chars as are in the input
+      // we only advance f as a last step after the other checks
+      // this is in case its a forward iterator
+      if (f == l) {
+        break;
+      }
+    }
+    auto begin = std::begin(data);
+    // from chars doesn't accept leading plus
+    if ( *begin == '+' ) {
+      ++begin;
+    }
+    auto out = attribute{};
+    const auto [ptr, ec] = std::from_chars(begin, end, out, Radix);
+    if (ec == std::errc{}) {
+      const auto read_chars = ptr - std::begin(data);
+      TENZIR_ASSERT(read_chars == take_chars);
+      if (read_chars >= MinDigits + first_is_sign) {
+        a = out;
+        return true;
+      }
+    }
+    f = save; // this doesnt actually work for forward/stream iterators
+    //TODO no parser that employs the `f = save` pattern actually works for these
+    // it would be a good idea to guard all parsers against this.
     return false;
```

Comment:
`Forward Iterators` do indeed provide multi-pass guarantees, but `Input Iterators` do not.

We had this issue with the stream parser test (that was disabled instead of removed), where you cannot revert the iterator to its starting position (or rather the staring position iterator is invalidated by advancing it).

It *would* be prudent to guard every parser reliant of `f = save;` against this, but that seems a bit out of scope.

Reply:
I think you got it the wrong way around. Forward iterators have multi-pass guarantee, input iterators do not, see https://en.cppreference.com/w/cpp/iterator/forward_iterator. 

But yes, in principle, I agree. Especially considering that we are moving away from this parsing framework, it does not seem worth it now though.